File: src/field_location.cpp
// field_location.cpp - classifies every scoring object into goals / loaders /
// the open field and publishes the GoalArray consumed by the scorer. Also
// turns controller buttons into intake / placement / toggle-flip calls.
#include <algorithm>
#include <cmath>
#include <cstdint>
#include <functional>
#include <memory>
#include <string>
#include <utility>
#include <vector>

#include <rclcpp/rclcpp.hpp>
#include <geometry_msgs/msg/pose_array.hpp>
#include <sensor_msgs/msg/joy.hpp>
#include <std_msgs/msg/int64_multi_array.hpp>

#include "override_sim/field_constants.h"
#include "override_sim/msg/cup_element.hpp"
#include "override_sim/msg/field_element.hpp"
#include "override_sim/msg/field_element_array.hpp"
#include "override_sim/msg/goal_array.hpp"
#include "override_sim/msg/goal_state.hpp"
#include "override_sim/msg/pin_element.hpp"
#include "override_sim/msg/toggle_array.hpp"
#include "override_sim/srv/flip_toggle.hpp"
#include "override_sim/srv/intake_element.hpp"
#include "override_sim/srv/score_element.hpp"

namespace override_sim
{

constexpr double kGoalXyTol = 0.13;       // horizontal distance that counts
constexpr double kLoaderXyTol = 0.14;
constexpr double kGoalEngageTol = 0.02;   // base may sink this much below rim
constexpr double kGoalMaxHeight = 0.35;   // base may be this high above rim

class FieldLocation : public rclcpp::Node
{
public:
  FieldLocation() : Node("field_location")
  {
    joy_sub_ = this->create_subscription<sensor_msgs::msg::Joy>(
      "/joy", 10,
      [this](const sensor_msgs::msg::Joy::SharedPtr msg) { controller_cb(msg); });
    elements_sub_ = this->create_subscription<FieldElementArray>(
      "/_object_locations", 10,
      [this](const FieldElementArray::SharedPtr msg) { object_location_cb(msg); });
    robot_pose_sub_ = this->create_subscription<geometry_msgs::msg::PoseArray>(
      "/otto_pose", 10,
      [this](const geometry_msgs::msg::PoseArray::SharedPtr msg) { robot_pose_cb(msg); });
    toggles_sub_ = this->create_subscription<ToggleArray>(
      "/toggles", 10,
      [this](const ToggleArray::SharedPtr msg) { toggle_cb(msg); });

    goals_pub_ = this->create_publisher<GoalArray>("/goals", 10);
    loaders_pub_ = this->create_publisher<std_msgs::msg::Int64MultiArray>(
      "/loaders", 10);
    field_objects_pub_ = this->create_publisher<FieldElementArray>(
      "/field_objects", 10);

    intake_client_ = this->create_client<IntakeElement>("/robot_intake");
    score_client_ = this->create_client<ScoreElement>("/score_element");
    flip_client_ = this->create_client<FlipToggle>("/flip_toggle");

    prev_buttons_ = {0, 0, 0, 0};
    toggle_states_ = {{0, 0}, {1, 0}, {2, 0}, {3, 0}};
  }

private:
  // ------------------------------------------------------------------
  // callbacks
  // ------------------------------------------------------------------
  void controller_cb(const sensor_msgs::msg::Joy::SharedPtr msg)
  {
    std::vector<int32_t> buttons;
    for (size_t i = 0; i < std::min<size_t>(msg->buttons.size(), 4); ++i) {
      buttons.push_back(msg->buttons[i]);
    }
    while (buttons.size() < 4) {
      buttons.push_back(0);
    }
    if (buttons[0] == 1 && prev_buttons_[0] == 0) {
      check_collision();                       // intake
    } else if (buttons[1] == 1 && prev_buttons_[1] == 0) {
      scoring_cb(1);                           // place pin
    } else if (buttons[2] == 1 && prev_buttons_[2] == 0) {
      scoring_cb(2);                           // place cup
    } else if (buttons[3] == 1 && prev_buttons_[3] == 0) {
      flip_nearest_toggle();                   // flip toggle
    }
    prev_buttons_ = buttons;
  }

  void robot_pose_cb(const geometry_msgs::msg::PoseArray::SharedPtr msg)
  {
    if (msg->poses.empty()) {
      return;
    }
    const auto & pose = msg->poses.back();
    robot_x_ = pose.position.x;
    robot_y_ = pose.position.y;
    robot_r_ = YawFromQuat(
      pose.orientation.x, pose.orientation.y, pose.orientation.z,
      pose.orientation.w);
  }

  void toggle_cb(const ToggleArray::SharedPtr msg)
  {
    for (const auto & t : msg->toggles) {
      toggle_states_[t.toggle_id] = t.state;
    }
  }

  void object_location_cb(const FieldElementArray::SharedPtr msg)
  {
    elements_ = msg;

    std::map<int, std::vector<FieldElement>> goal_stacks;
    std::map<int, std::vector<FieldElement>> loader_contents;
    for (const auto & kv : kGoals) {
      goal_stacks[kv.first] = {};
    }
    for (const auto & kv : kLoaders) {
      loader_contents[kv.first] = {};
    }
    auto field_elems = std::make_unique<FieldElementArray>();

    for (const auto & el : msg->elements) {
      const double base_z = el.location.z;
      const double x = el.location.x;
      const double y = el.location.y;
      bool placed = false;

      for (const auto & kv : kGoals) {
        const int gid = kv.first;
        const GoalInfo & g = kv.second;
        if (std::hypot(x - g.x, y - g.y) < kGoalXyTol &&
          g.height - kGoalEngageTol <= base_z &&
          base_z <= g.height + kGoalMaxHeight)
        {
          goal_stacks[gid].push_back(el);
          placed = true;
          break;
        }
      }
      if (placed) {
        continue;
      }

      for (const auto & kv : kLoaders) {
        const int lid = kv.first;
        const auto & p = kv.second;
        if (std::abs(x - p.first) < kLoaderXyTol &&
          std::abs(y - p.second) < kLoaderXyTol && base_z < 0.4)
        {
          loader_contents[lid].push_back(el);
          placed = true;
          break;
        }
      }

      if (!placed) {
        field_elems->elements.push_back(el);
      }
    }

    // build the GoalArray (kGoals is ordered, so ids come out sorted)
    auto goal_array = std::make_unique<GoalArray>();
    for (const auto & kv : kGoals) {
      const int gid = kv.first;
      const GoalInfo & g = kv.second;
      std::vector<FieldElement> stack = goal_stacks[gid];
      std::sort(stack.begin(), stack.end(),
        [](const FieldElement & a, const FieldElement & b) {
          return a.location.z < b.location.z;
        });

      GoalState gs;
      gs.goal_id = gid;
      gs.goal_type = static_cast<int8_t>(g.type);
      gs.quadrant = g.quadrant;
      gs.toggle_state = static_cast<int8_t>(
        toggle_states_.at(QuadrantToToggle(g.quadrant)));

      for (const auto & el : stack) {
        if (el.element_type == 1) {
          PinElement pin;
          pin.object_name = el.object_name;
          pin.id = el.id;
          pin.top_color = el.top_color;
          pin.bottom_color = el.bottom_color;
          pin.location = el.location;
          pin.orientation = el.orientation;
          gs.pins.push_back(pin);
        } else {
          CupElement cup;
          cup.object_name = el.object_name;
          cup.id = el.id;
          cup.location = el.location;
          cup.orientation = el.orientation;
          cup.yaw = YawFromQuat(
            el.orientation.x, el.orientation.y, el.orientation.z,
            el.orientation.w);
          gs.cups.push_back(cup);
        }
      }
      goal_array->goals.push_back(gs);
    }

    goals_pub_->publish(std::move(goal_array));

    auto loader_msg = std::make_unique<std_msgs::msg::Int64MultiArray>();
    for (const auto & kv : kLoaders) {
      loader_msg->data.push_back(
        static_cast<int64_t>(loader_contents[kv.first].size()));
    }
    loaders_pub_->publish(std::move(loader_msg));

    field_objects_pub_->publish(std::move(field_elems));
  }

  // ------------------------------------------------------------------
  // helpers
  // ------------------------------------------------------------------
  void check_collision()
  {
    // is there an element in the robot's intake zone?
    const double h = robot_x_;
    const double k = robot_y_;
    const double th = robot_r_;
    constexpr double offset = 0.15;
    const double ref_x = h + offset * std::cos(th);
    const double ref_y = k + offset * std::sin(th);

    for (const auto & el : elements_->elements) {
      const double dx = el.location.x - ref_x;
      const double dy = el.location.y - ref_y;
      const double dx_r = std::cos(th) * dx + std::sin(th) * dy;
      const double dy_r = -std::sin(th) * dx + std::cos(th) * dy;
      if (std::abs(dx_r) < 0.10 && std::abs(dy_r) < 0.10 &&
        el.location.z < 0.12)
      {
        auto req = std::make_shared<IntakeElement::Request>();
        req->entity_id = el.id;
        req->entity_type = el.element_type;
        req->top_color = el.top_color;
        req->bottom_color = el.bottom_color;
        intake_client_->async_send_request(req);
        RCLCPP_INFO(this->get_logger(),
          "intaking %s (type %d)", el.object_name.c_str(), el.element_type);
        return;
      }
    }
    RCLCPP_INFO(this->get_logger(), "no element in intake zone");
  }

  int check_location()
  {
    int best = 0;
    double best_d = 0.7;
    for (const auto & kv : kGoals) {
      const double d = std::hypot(robot_x_ - kv.second.x, robot_y_ - kv.second.y);
      if (d < best_d) {
        best = kv.first;
        best_d = d;
      }
    }
    return best;
  }

  void scoring_cb(int8_t element_type)
  {
    const int goal_id = check_location();
    RCLCPP_INFO(this->get_logger(), "placing element type %d at goal %d",
      element_type, goal_id);
    auto req = std::make_shared<ScoreElement::Request>();
    req->element_type = element_type;
    req->top_color = 0;
    req->bottom_color = 0;
    req->goal_id = goal_id;
    score_client_->async_send_request(req);
  }

  void flip_nearest_toggle()
  {
    int best = -1;
    double best_d = 1.0;
    for (const auto & kv : kToggles) {
      const auto & t = kv.second;
      const double d = std::hypot(robot_x_ - t[0], robot_y_ - t[1]);
      if (d < best_d) {
        best = kv.first;
        best_d = d;
      }
    }
    if (best == -1) {
      RCLCPP_INFO(this->get_logger(), "no toggle within range");
      return;
    }
    auto req = std::make_shared<FlipToggle::Request>();
    req->toggle_id = best;
    req->state = -1;
    flip_client_->async_send_request(req);
    RCLCPP_INFO(this->get_logger(), "flipping toggle %d", best);
  }

  // state
  FieldElementArray::SharedPtr elements_;
  std::vector<int32_t> prev_buttons_;
  std::map<int, int8_t> toggle_states_;
  double robot_x_ = 0.0;
  double robot_y_ = 0.0;
  double robot_r_ = 0.0;

  rclcpp::Subscription<sensor_msgs::msg::Joy>::SharedPtr joy_sub_;
  rclcpp::Subscription<FieldElementArray>::SharedPtr elements_sub_;
  rclcpp::Subscription<geometry_msgs::msg::PoseArray>::SharedPtr robot_pose_sub_;
  rclcpp::Subscription<ToggleArray>::SharedPtr toggles_sub_;
  rclcpp::Publisher<GoalArray>::SharedPtr goals_pub_;
  rclcpp::Publisher<std_msgs::msg::Int64MultiArray>::SharedPtr loaders_pub_;
  rclcpp::Publisher<FieldElementArray>::SharedPtr field_objects_pub_;
  rclcpp::Client<IntakeElement>::SharedPtr intake_client_;
  rclcpp::Client<ScoreElement>::SharedPtr score_client_;
  rclcpp::Client<FlipToggle>::SharedPtr flip_client_;
};

}  // namespace override_sim

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<override_sim::FieldLocation>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}

File: src/field_location.py
#!/usr/bin/env python3

import math

import rclpy
import numpy as np
from rclpy.node import Node
from scipy.spatial.transform import Rotation as R

from sensor_msgs.msg import Joy
from geometry_msgs.msg import PoseArray
from std_msgs.msg import Int64MultiArray

from override_sim.msg import (
    CupElement, FieldElement, FieldElementArray, GoalArray, GoalState,
    PinElement, ToggleArray,
)
from override_sim.srv import FlipToggle, IntakeElement, ScoreElement

# ---------------------------------------------------------------------------
# Override field reference data (meters, field center at origin)
# ---------------------------------------------------------------------------
# goal_id: (x, y, type, quadrant, height)
# type: 0 = tall neutral center, 1 = short neutral, 2 = red alliance, 3 = blue
GOALS = {
    0: (0.0, 0.0, 0, 'C', 0.2227),
    1: (-0.6, 1.2, 1, 'N', 0.1465),
    2: (-1.2, 0.6, 1, 'W', 0.1465),
    3: (-1.2, -0.6, 2, 'W', 0.0825),
    4: (-0.6, -1.2, 2, 'S', 0.0825),
    5: (0.6, 1.2, 3, 'N', 0.0825),
    6: (1.2, 0.6, 3, 'E', 0.0825),
    7: (1.2, -0.6, 1, 'E', 0.1465),
    8: (0.6, -1.2, 1, 'S', 0.1465),
}

TOGGLE_QUADRANTS = {0: 'N', 1: 'E', 2: 'S', 3: 'W'}

LOADERS = {
    0: (-1.74, 1.49),   # red station, top-left
    1: (-1.74, -1.49),  # red station, bottom-left
    2: (1.74, 1.49),    # blue station, top-right
    3: (1.74, -1.49),   # blue station, bottom-right
}

GOAL_XY_TOL = 0.13       # horizontal distance from goal center that counts
LOADER_XY_TOL = 0.14
GOAL_ENGAGE_TOL = 0.02   # base may sink this much below the goal rim
GOAL_MAX_HEIGHT = 0.35   # element base may be this high above the rim


class FieldLocation(Node):
    """Classifies every scoring object into goals / loaders / field and
    publishes the resulting GoalArray. Also turns controller buttons into
    intake / placement / toggle-flip service calls."""

    def __init__(self):
        super().__init__('field_location')

        # teleop controller feedback
        self.create_subscription(Joy, '/joy', self.controller_callback, 10)
        # element poses from pose_bridge.py
        self.create_subscription(
            FieldElementArray, '/_object_locations', self.object_location_callback, 10)
        # robot pose
        self.create_subscription(PoseArray, '/otto_pose', self.robot_pose_callback, 10)
        # toggle states from world_services.py
        self.create_subscription(ToggleArray, '/toggles', self.toggle_callback, 10)

        # publishers
        self.goals = self.create_publisher(GoalArray, '/goals', 10)
        self.loaders = self.create_publisher(Int64MultiArray, '/loaders', 10)
        self.field_objects = self.create_publisher(
            FieldElementArray, '/field_objects', 10)

        # service clients
        self.intake_element = self.create_client(IntakeElement, '/robot_intake')
        self.score_element = self.create_client(ScoreElement, '/score_element')
        self.flip_toggle = self.create_client(FlipToggle, '/flip_toggle')

        # controller debounce
        self.prev_buttons = [0] * 4

        self.toggle_states = {0: 0, 1: 0, 2: 0, 3: 0}
        self.elements = FieldElementArray()

    # ------------------------------------------------------------------
    # callbacks
    # ------------------------------------------------------------------
    def controller_callback(self, msg: Joy):
        buttons = list(msg.buttons[:4]) if len(msg.buttons) >= 4 else list(msg.buttons)
        if buttons[0] == 1 and self.prev_buttons[0] == 0:      # intake
            self.check_collision()
        elif buttons[1] == 1 and self.prev_buttons[1] == 0:    # place pin
            self.scoring_callback(1)
        elif buttons[2] == 1 and self.prev_buttons[2] == 0:    # place cup
            self.scoring_callback(2)
        elif buttons[3] == 1 and self.prev_buttons[3] == 0:    # flip toggle
            self.flip_nearest_toggle()
        self.prev_buttons = buttons

    def robot_pose_callback(self, msg: PoseArray):
        self.robot_x = msg.poses[-1].position.x
        self.robot_y = msg.poses[-1].position.y
        quat = msg.poses[-1].orientation
        q = np.array([quat.x, quat.y, quat.z, quat.w])
        q = q / (np.linalg.norm(q) or 1.0)
        self.robot_r = float(R.from_quat(q).as_euler('xyz')[2])

    def toggle_callback(self, msg: ToggleArray):
        for t in msg.toggles:
            self.toggle_states[t.toggle_id] = t.state

    def object_location_callback(self, msg: FieldElementArray):
        self.elements = msg

        goal_stacks = {gid: [] for gid in GOALS}
        loader_contents = {lid: [] for lid in LOADERS}
        field_elems = FieldElementArray()

        for el in msg.elements:
            base_z = el.location.z  # model origin == element base
            x, y = el.location.x, el.location.y
            placed = False

            # check goals (lowest id wins to keep assignment deterministic)
            for gid, (gx, gy, _gtype, _gquad, gheight) in GOALS.items():
                if math.hypot(x - gx, y - gy) < GOAL_XY_TOL:
                    if gheight - GOAL_ENGAGE_TOL <= base_z <= gheight + GOAL_MAX_HEIGHT:
                        goal_stacks[gid].append(el)
                        placed = True
                        break

            if placed:
                continue

            # check loaders
            for lid, (lx, ly) in LOADERS.items():
                if (abs(x - lx) < LOADER_XY_TOL and abs(y - ly) < LOADER_XY_TOL
                        and base_z < 0.4):
                    loader_contents[lid].append(el)
                    placed = True
                    break

            if not placed:
                field_elems.elements.append(el)

        # build the GoalArray
        goal_array = GoalArray()
        for gid in sorted(GOALS):
            gx, gy, gtype, gquad, gheight = GOALS[gid]
            stack = sorted(goal_stacks[gid], key=lambda e: e.location.z)
            gs = GoalState()
            gs.goal_id = gid
            gs.goal_type = gtype
            gs.quadrant = gquad
            gs.toggle_state = self.toggle_states.get(
                self.quadrant_to_toggle(gquad), 0)
            for el in stack:
                if el.element_type == 1:
                    pin = PinElement()
                    pin.object_name = el.object_name
                    pin.id = el.id
                    pin.top_color = el.top_color
                    pin.bottom_color = el.bottom_color
                    pin.location = el.location
                    pin.orientation = el.orientation
                    gs.pins.append(pin)
                else:
                    cup = CupElement()
                    cup.object_name = el.object_name
                    cup.id = el.id
                    cup.location = el.location
                    cup.orientation = el.orientation
                    cup.yaw = self.quat_to_yaw(el.orientation)
                    gs.cups.append(cup)
            goal_array.goals.append(gs)

        # publish
        self.goals.publish(goal_array)

        loader_msg = Int64MultiArray()
        loader_msg.data = [len(loader_contents[i]) for i in sorted(LOADERS)]
        self.loaders.publish(loader_msg)

        self.field_objects.publish(field_elems)

    # ------------------------------------------------------------------
    # helpers
    # ------------------------------------------------------------------
    @staticmethod
    def quat_to_yaw(orientation):
        q = np.array([orientation.x, orientation.y, orientation.z, orientation.w])
        q = q / (np.linalg.norm(q) or 1.0)
        return float(R.from_quat(q).as_euler('xyz')[2])

    @staticmethod
    def quadrant_to_toggle(quadrant):
        for tid, q in TOGGLE_QUADRANTS.items():
            if q == quadrant:
                return tid
        return -1

    def check_collision(self):
        """Is there a scoring element in the robot's intake zone?"""
        h, k, th = self.robot_x, self.robot_y, self.robot_r
        offset = 0.15
        ref_x = h + offset * math.cos(th)
        ref_y = k + offset * math.sin(th)

        for el in self.elements.elements:
            dx, dy = el.location.x - ref_x, el.location.y - ref_y
            dx_r = math.cos(th) * dx + math.sin(th) * dy
            dy_r = -math.sin(th) * dx + math.cos(th) * dy
            if abs(dx_r) < 0.10 and abs(dy_r) < 0.10 and el.location.z < 0.12:
                req = IntakeElement.Request()
                req.entity_id = el.id
                req.entity_type = el.element_type
                req.top_color = el.top_color
                req.bottom_color = el.bottom_color
                self.intake_element.call_async(req)
                self.get_logger().info(
                    f"intaking {el.object_name} (type {el.element_type})")
                return
        self.get_logger().info("no element in intake zone")

    def check_location(self):
        """Nearest goal id to the robot (0 if too far)."""
        best, best_d = 0, 0.7
        for gid, (gx, gy, *_rest) in GOALS.items():
            d = math.hypot(self.robot_x - gx, self.robot_y - gy)
            if d < best_d:
                best, best_d = gid, d
        return best

    def scoring_callback(self, element_type):
        goal_id = self.check_location()
        self.get_logger().info(f'placing element type {element_type} at goal {goal_id}')
        req = ScoreElement.Request()
        req.element_type = element_type
        req.top_color = 0
        req.bottom_color = 0
        req.goal_id = goal_id
        self.score_element.call_async(req)

    def flip_nearest_toggle(self):
        best, best_d = -1, 1.0
        for tid, quadrant in TOGGLE_QUADRANTS.items():
            tx, ty = self.toggle_position(tid)
            d = math.hypot(self.robot_x - tx, self.robot_y - ty)
            if d < best_d:
                best, best_d = tid, d
        if best == -1:
            self.get_logger().info("no toggle within range")
            return
        req = FlipToggle.Request()
        req.toggle_id = best
        req.state = -1
        self.flip_toggle.call_async(req)
        self.get_logger().info(f"flipping toggle {best}")

    @staticmethod
    def toggle_position(toggle_id):
        quadrant = TOGGLE_QUADRANTS[toggle_id]
        if quadrant == 'N':
            return (0.0, 1.78)
        if quadrant == 'E':
            return (1.78, 0.0)
        if quadrant == 'S':
            return (0.0, -1.78)
        return (-1.78, 0.0)


def main(args=None):
    rclpy.init(args=args)
    node = FieldLocation()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == '__main__':
    main()
